parser/v2/goexpression/parse.go

```diff
@@ -0,0 +1,257 @@
+package goexpression
+
+import (
+	"errors"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"strings"
+	"unicode"
+)
+
+var ErrContainerFuncNotFound = errors.New("parser error: templ container function not found")
+var ErrExpectedNodeNotFound = errors.New("parser error: expected node not found")
+
+func Case(content string) (start, end int, err error) {
+	if !(strings.HasPrefix(content, "case") || strings.HasPrefix(content, "default")) {
+		return 0, 0, ErrExpectedNodeNotFound
+	}
+	prefix := "switch {\n"
+	start, end, err = extract(prefix+content+"\n}", CaseExtractor{})
+	if err != nil {
+		return 0, 0, err
+	}
+	// Since we added a `switch {` prefix, we need to remove it.
+	start -= len(prefix)
+	end -= len(prefix)
+	return start, end, nil
+}
+
+func If(content string) (start, end int, err error) {
+	if !strings.HasPrefix(content, "if") {
+		return 0, 0, ErrExpectedNodeNotFound
+	}
+	return extract(content, IfExtractor{})
+}
+
+func For(content string) (start, end int, err error) {
+	if !strings.HasPrefix(content, "for") {
+		return 0, 0, ErrExpectedNodeNotFound
+	}
+	return extract(content, ForExtractor{})
+}
+
+func Switch(content string) (start, end int, err error) {
+	if !strings.HasPrefix(content, "switch") {
+		return 0, 0, ErrExpectedNodeNotFound
+	}
+	return extract(content, SwitchExtractor{})
+}
+
+func Expression(content string) (start, end int, err error) {
+	start, end, err = extract(content, ExprExtractor{})
+	if err != nil {
+		return 0, 0, err
+	}
+	// If the expression ends with `...` then it's a child spread expression.
+	if suffix := content[end:]; strings.HasPrefix(suffix, "...") {
+		end += len("...")
+	}
+	return start, end, nil
+}
+
+type IfExtractor struct{}
+
+var InGoIfExpression = false
+
+func (e IfExtractor) Code(src string, body []ast.Stmt) (start, end int, err error) {
+	stmt, ok := body[0].(*ast.IfStmt)
+	if !ok {
+		return 0, 0, ErrExpectedNodeNotFound
+	}
+	start = int(stmt.If) + 2
+	end = latestEnd(start, stmt.Init, stmt.Cond)
+	return start, end, nil
+}
+
+type ForExtractor struct{}
+
+func latestEnd(start int, nodes ...ast.Node) (end int) {
```

Comment:
Nit: This function could be better placed above or below all the extractors.

Reply:
Good idea. Looking at it again, I've decided to remove the extractor types completely and inline the functions. It increases the locality of behaviour which I think makes it easier it reader.